typedb/localstack_typedb/utils/docker.py

```diff
@@ -0,0 +1,197 @@
+import re
+import logging
+from functools import cache
+from typing import Callable
+import requests
+
+from localstack import config
+from localstack.config import is_env_true
+from localstack_typedb.utils.h2_proxy import (
+    apply_http2_patches_for_grpc_support,
+    ProxyRequestMatcher,
+)
+from localstack.utils.docker_utils import DOCKER_CLIENT
+from localstack.extensions.api import Extension, http
+from localstack.http import Request
+from localstack.utils.container_utils.container_client import PortMappings
+from localstack.utils.net import get_addressable_container_host
+from localstack.utils.sync import retry
+from rolo import route
+from rolo.proxy import Proxy
+from rolo.routing import RuleAdapter, WithHost
+
+LOG = logging.getLogger(__name__)
+logging.getLogger("localstack_typedb").setLevel(
+    logging.DEBUG if config.DEBUG else logging.INFO
+)
+logging.basicConfig()
+
+
+class ProxiedDockerContainerExtension(Extension, ProxyRequestMatcher):
+    """
+    Utility class to create a LocalStack Extension backed by a Docker container that exposes a service
+    on a network port (or several ports), with requests being proxied through the LocalStack gateway.
+
+    Requests may potentially use HTTP2 with binary content as the protocol (e.g., gRPC over HTTP2).
+    To ensure proper routing of requests, subclasses can define the `http2_ports`.
+    """
+
+    name: str
+    """Name of this extension"""
+    image_name: str
+    """Docker image name"""
+    container_name: str | None
+    """Name of the Docker container spun up by the extension"""
+    container_ports: list[int]
+    """List of network ports of the Docker container spun up by the extension"""
+    host: str | None
+    """
+    Optional host on which to expose the container endpoints.
+    Can be either a static hostname, or a pattern like `<regex("(.+\.)?"):subdomain>myext.<domain>`
+    """
+    path: str | None
+    """Optional path on which to expose the container endpoints."""
+    command: list[str] | None
+    """Optional command (and flags) to execute in the container."""
+
+    request_to_port_router: Callable[[Request], int] | None
+    """Callable that returns the target port for a given request, for routing purposes"""
+    http2_ports: list[int] | None
+    """List of ports for which HTTP2 proxy forwarding into the container should be enabled."""
+
+    def __init__(
+        self,
+        image_name: str,
+        container_ports: list[int],
+        host: str | None = None,
+        path: str | None = None,
+        container_name: str | None = None,
+        command: list[str] | None = None,
+        request_to_port_router: Callable[[Request], int] | None = None,
+        http2_ports: list[int] | None = None,
+    ):
+        self.image_name = image_name
+        self.container_ports = container_ports
```

Comment:
Nice to fail fast in the case that an empty list is passed, since the type checker can't detect it, and `self.container_ports[0]` would explode later.
```suggestion
        if not container_ports:
            raise ArgumentError("container_ports cannot be empty")
        self.container_ports = container_ports
```